### PR TITLE
JWT ENABLED JMSPUT

### PR DESCRIPTION
**Pull request by harshitsharma2801**
1. Modified env.json with JWT_ENDPOINTS.
2. Changes in SampleEnvSetter to populate the jwtEndpoints.
3. Added JwtHelper class for all http request related code.
4. Changes in JmsPut to handle Jwt_endpoints.

### CODE REVIEW

File: JMS/com/ibm/mq/samples/jms/JmsPut.java [hunk 1]
```diff
@@ -170,11 +179,17 @@ private static void setJMSProperties(JmsConnectionFactory cf) {
...
-            if (null != APP_USER && !APP_USER.trim().isEmpty()) {
-                cf.setBooleanProperty(WMQConstants.USER_AUTHENTICATION_MQCSP, true);
-                cf.setStringProperty(WMQConstants.USERID, APP_USER);
-                cf.setStringProperty(WMQConstants.PASSWORD, APP_PASSWORD);
+            cf.setBooleanProperty(WMQConstants.USER_AUTHENTICATION_MQCSP, true);
+            if(accessToken != null){
+                cf.setStringProperty(WMQConstants.USERID, null);
+                cf.setStringProperty(WMQConstants.PASSWORD, accessToken);
+            }else{
+                if(null != APP_USER && !APP_USER.trim().isEmpty()){
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/JmsPut.java:187]: house style

File: JMS/com/ibm/mq/samples/jms/JmsPut.java [hunk 2]
```diff
@@ -170,11 +179,17 @@ private static void setJMSProperties(JmsConnectionFactory cf) {
 
             cf.setStringProperty(WMQConstants.WMQ_QUEUE_MANAGER, QMGR);
             cf.setStringProperty(WMQConstants.WMQ_APPLICATIONNAME, APP_NAME);
-            if (null != APP_USER && !APP_USER.trim().isEmpty()) {
-                cf.setBooleanProperty(WMQConstants.USER_AUTHENTICATION_MQCSP, true);
-                cf.setStringProperty(WMQConstants.USERID, APP_USER);
-                cf.setStringProperty(WMQConstants.PASSWORD, APP_PASSWORD);
+            cf.setBooleanProperty(WMQConstants.USER_AUTHENTICATION_MQCSP, true);
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/JmsPut.java:182]: This logic block is large enough to warrant its own private method.

File: JMS/com/ibm/mq/samples/jms/JmsPut.java [hunk 3]
```diff
@@ -170,11 +179,17 @@ private static void setJMSProperties(JmsConnectionFactory cf) {
 
             cf.setStringProperty(WMQConstants.WMQ_QUEUE_MANAGER, QMGR);
             cf.setStringProperty(WMQConstants.WMQ_APPLICATIONNAME, APP_NAME);
-            if (null != APP_USER && !APP_USER.trim().isEmpty()) {
-                cf.setBooleanProperty(WMQConstants.USER_AUTHENTICATION_MQCSP, true);
-                cf.setStringProperty(WMQConstants.USERID, APP_USER);
-                cf.setStringProperty(WMQConstants.PASSWORD, APP_PASSWORD);
+            cf.setBooleanProperty(WMQConstants.USER_AUTHENTICATION_MQCSP, true);
+            if(accessToken != null){
+                cf.setStringProperty(WMQConstants.USERID, null);
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/JmsPut.java:184]: does it need to be set to null? can it be not set at all to work?

File: JMS/com/ibm/mq/samples/jms/JwtHelper.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import java.net.http.HttpRequest;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.util.logging.Logger;
+import java.net.http.HttpResponse;
+import org.json.JSONObject;
+import com.ibm.mq.samples.jms.SampleEnvSetter;
+
+public class JwtHelper {
+    private static final Logger logger = Logger.getLogger("com.ibm.mq.samples.jms");
+    private static String tokenEndpoint = "";
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/JwtHelper.java:29]: why static properties, why not instance properties?

File: JMS/com/ibm/mq/samples/jms/JwtHelper.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        } catch (Exception e) {
+          logger.info("Using token:" + access_token);
+          e.printStackTrace();
+        }
+        return access_token;
+    }
+
+    public boolean isJwtEnabled(){
+
+        if((tokenEndpoint == null || tokenEndpoint.isEmpty()) ||
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/JwtHelper.java:76]: should trim(), before checking for isEmtpy. ie
`tokenEndpoint.trim().isEmpty()`

File: JMS/com/ibm/mq/samples/jms/SampleEnvSetter.java [hunk 6]
```diff
@@ -61,13 +61,19 @@ public SampleEnvSetter() {
             if (mqEnvSettings != null) {
               logger.info("JSON Data Found");
               mqEndPoints = (JSONArray) mqEnvSettings.getJSONArray("MQ_ENDPOINTS");
+              if(mqEnvSettings.has("JWT_ISSUER") ){
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/SampleEnvSetter.java:64]: house style

File: JMS/com/ibm/mq/samples/jms/SampleEnvSetter.java [hunk 7]
```diff
@@ -233,4 +239,23 @@ public int getCount() {
         // need to come from a CCDT and environment settings
         return (null == mqEndPoints) ? 1 : mqEndPoints.length();
     }
+
+    public String getJwtEnv(String key){
+        String value = System.getProperty(key);
+        try {
+            if((value == null || value.isEmpty()) &&
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/SampleEnvSetter.java:246]: house style

**Review by chughts — CHANGES_REQUESTED** [review]

File: JMS/com/ibm/mq/samples/jms/JwtHelper.java (hunk 4 above)
Comment (harshitsharma2801) [reply JMS/com/ibm/mq/samples/jms/JwtHelper.java:29]: These properties are expected to remain same for every instance of JwtHelper.

File: JMS/com/ibm/mq/samples/jms/JmsPut.java [hunk 8]
```diff
@@ -68,10 +69,18 @@ public class JmsPut {
...
+    private static String accessToken = null;
     public static void main(String[] args) {
         initialiseLogging();
-        mqConnectionVariables();
+        SampleEnvSetter env = new SampleEnvSetter();
+        jh = new JwtHelper(env);
+        if (jh.isJwtEnabled()) {
+            accessToken = jh.obtainToken();
+        } else {
+            logger.info("One or more JWT Credentials missing! Expected APP_USER and APP_PASSWORD in MQ Endpoints");
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/JmsPut.java:81]: It is now an info message - good, but the second clause is alarming. It should be more like "One of more JWT Credentials missing! Will not be using JWT for authentication"

File: JMS/com/ibm/mq/samples/jms/JwtHelper.java (hunk 4 above)
Comment (chughts) [reply JMS/com/ibm/mq/samples/jms/JwtHelper.java:29]: In that case they should be set once, in the first time a constructor is called, protected by an if ! empty block so they only get set once, and all instances of JwtHelper use them.

File: JMS/com/ibm/mq/samples/jms/SampleEnvSetter.java [hunk 9]
```diff
@@ -61,13 +61,19 @@ public SampleEnvSetter() {
             if (mqEnvSettings != null) {
               logger.info("JSON Data Found");
               mqEndPoints = (JSONArray) mqEnvSettings.getJSONArray("MQ_ENDPOINTS");
+              if (mqEnvSettings.has("JWT_ISSUER")) {
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/SampleEnvSetter.java:64]: I think this is ok, but the clause becomes `if (mqEnvSettings != null) && mqEnvSettings.has("JWT_ISSUER"))` 

It makes sense to put this outside the `if (mqEnvSettings != null)`

File: JMS/com/ibm/mq/samples/jms/SampleEnvSetter.java [hunk 10]
```diff
@@ -61,13 +61,19 @@ public SampleEnvSetter() {
...
+                  jwtEndPoints = (JSONObject) mqEnvSettings.getJSONObject("JWT_ISSUER");
+              }
             }
 
             if (mqEndPoints == null || mqEndPoints.isEmpty()) {
                 logger.warning("No Endpoints found in .json file next instruction " +
                                  "will raise a null pointer exception");
             } else {
                 logger.info("There is at least one MQ endpoint in the .json file");
+                if (jwtEndPoints != null) {
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/SampleEnvSetter.java:74]: Again doesn't need to be inside the `if (mqEndPoints == null)` clause.

**Review by chughts — CHANGES_REQUESTED** [review]

File: JMS/com/ibm/mq/samples/jms/JmsPut.java [hunk 11]
```diff
@@ -68,10 +69,18 @@ public class JmsPut {
     private static String CIPHER_SUITE;
     private static String CCDTURL;
     private static Boolean BINDINGS = false;
-
+    private static JwtHelper jh = null;
+    private static String accessToken = null;
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/JmsPut.java:73]: add line separation before main definition, otherwise it gets lost.

**Review by chughts — CHANGES_REQUESTED** [review]

File: JMS/com/ibm/mq/samples/jms/JwtHelper.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+public class JwtHelper {
+    private static final Logger logger = Logger.getLogger("com.ibm.mq.samples.jms");
+    private static String tokenEndpoint = "";
+    private static String tokenUsername = "";
+    private static String tokenPassword = "";
+    private static String tokenClientId = "";
+    
+    public JwtHelper(SampleEnvSetter env) {
+        if (tokenEndpoint.trim().isEmpty()) {
```
Comment (chughts) [inline JMS/com/ibm/mq/samples/jms/JwtHelper.java:35]: I was thinking of letting this go, but it sticks out too much. The normal way to flag whether static values have been set, is to see a static initialised boolean, which defaults to false. Then there is only one if block.

**Review by chughts — CHANGES_REQUESTED** [review]

**Review by chughts — APPROVED** [review]

**Pull request merged**